server/container_update_resources.go

```diff
@@ -15,6 +17,47 @@ import (
 	"github.com/cri-o/cri-o/internal/oci"
 )
 
+// validateMemoryUpdate checks if the new memory limit is safe to apply by getting current usage from CRI-O's stats server.
+// This ensures real-time accuracy and prevents decreasing memory limits below current usage.
+func (s *Server) validateMemoryUpdate(ctx context.Context, c *oci.Container, newMemoryLimit int64) error {
+	if newMemoryLimit <= 0 {
+		return nil
+	}
+
+	sb := s.GetSandbox(c.Sandbox())
+	if sb == nil {
+		log.Warnf(ctx, "Could not get sandbox %s for container %s", c.Sandbox(), c.ID())
+
+		return nil
+	}
+
+	containerStats := s.StatsForContainer(c, sb)
+	if containerStats == nil || containerStats.GetMemory() == nil {
+		log.Warnf(ctx, "No memory stats available for container %s", c.ID())
+
+		return nil
+	}
+
+	if containerStats.GetMemory().GetUsageBytes() == nil {
+		log.Warnf(ctx, "Memory usage not available for container %s", c.ID())
+
+		return nil
+	}
+
+	currentUsage := int64(containerStats.GetMemory().GetUsageBytes().GetValue())
```

Comment:
nil check is not required for auto-generated proto struct. Get*() chain is smart enough and if one of them is nil, the whole chain returns nil without panic.
We just need to check whether the final value is nil or not.
https://github.com/cri-o/cri-o/blob/ab3f85e5fc00b55ca036cb0b61252310db5d9b0f/vendor/k8s.io/cri-api/pkg/apis/runtime/v1/api.pb.go#L9353-L9358